core/test/utils/matrix_generator.hpp

```diff
@@ -163,52 +189,72 @@ std::unique_ptr<MatrixType> generate_random_matrix(
 template <typename ValueType, typename IndexType, typename NonzeroDistribution,
           typename ValueDistribution, typename Engine>
 matrix_data<ValueType, IndexType> generate_random_triangular_matrix_data(
-    size_type num_rows, size_type num_cols, bool ones_on_diagonal,
-    bool lower_triangular, NonzeroDistribution&& nonzero_dist,
-    ValueDistribution&& value_dist, Engine&& engine)
+    size_type size, bool ones_on_diagonal, bool lower_triangular,
+    NonzeroDistribution&& nonzero_dist, ValueDistribution&& value_dist,
+    Engine&& engine)
 {
     using std::begin;
     using std::end;
 
-    matrix_data<ValueType, IndexType> data{gko::dim<2>{num_rows, num_cols}, {}};
-    ValueType one = 1.0;
-    std::vector<size_type> col_idx(num_cols);
-    std::iota(begin(col_idx), end(col_idx), size_type(0));
+    matrix_data<ValueType, IndexType> data{gko::dim<2>{size, size}, {}};
+
+    std::vector<bool> present_cols(size);
 
-    for (size_type row = 0; row < num_rows; ++row) {
+    for (IndexType row = 0; row < size; ++row) {
         // randomly generate number of nonzeros in this row
-        auto nnz_in_row = static_cast<size_type>(nonzero_dist(engine));
-        nnz_in_row = std::max(size_type(0), std::min(nnz_in_row, num_cols));
-        // select a subset of `nnz_in_row` column indexes, and fill these
-        // locations with random values
-        std::shuffle(begin(col_idx), end(col_idx), engine);
-        // add non-zeros
-        bool has_diagonal{};
-        for (size_type nz = 0; nz < nnz_in_row; ++nz) {
-            auto col = col_idx[nz];
-            // skip non-zeros outside triangle
-            if ((col > row && lower_triangular) ||
-                (col < row && !lower_triangular)) {
```

Comment:
Is it not allowed to generate the rectangular matrix filling upper/lower part?

Reply:
if you want something like that, you can call the function and adjust the matrix_data::size afterwards. This doesn't seem worth the effort to me, since we don't use rectangular triangular matrices anywhere.

Reply:
no, you can not just adjust the matrix size. if generate the lower part in the tall and skinny matrix, it will also fill the value in the nonsquare part.

Reply:
sure, you only need to start with dimension `min(rows, cols)` and change it to `{rows, cols}` afterwards

Reply:
sorry, my mistake, I was not thinking of the trailing part. However, I still don't see a place where we need this.